docs/design-documents/statistics/network_stats.md

```diff
@@ -0,0 +1,130 @@
+
+## Data collection API
+
+This is proposal of the new API that provides network statistics. Some parts can be done in Mbed OS 5.11 timeframe without API changes. I'll also propose API changes for 5.12 to collect more data.
+
+### Socket status list
+
+We could show some kind of list of sockets that are currently open, or have been open. Instead of abstract ID, we provide pointers to the `Socket` objects, we those pointers are not thread safe, because the original object might already been desctructed when data is parsed. Therefore the last associated destination address is copied into this structure. Or if we want to save memory, we could record just the destination port number.
+
+```
+typedef enum socket_state {
+    CLOSED,                         /**< Socket is closed and does not exist anymore in the system */
+    OPEN,                           /**< Socket is open, but not associated to any peer address */
+    CONNECTED,                      /**< Socket is associated to peer address, either by connect() or sendto()/recvfrom() calls */
+    LISTEN,                         /**< Socket is listening for incoming connections */
+};
+
+typedef struct {
+    Socket *socket;                 /**< Pointer to original Socket objet. Not safe to access, used for identifying */
+    SocketAddress peername          /**< Last associated peername of this socket (Destination address) */
+    socket_state state;             /**< State of this socket */
+    uint32_t tick_last_change;      /**< osKernelGetTick() when state last changed */
+    size_t sent;                    /**< Data sent through this socket */
+    size_t received;                /**< Data received through this socket */
+} mbed_stats_socket_t
+
+/** Write socket statistics into structure pointed by stats.
+  * @count How many structures can be written
+  * @return How many structures was written
+  */
+size_t mbed_stats_socket_get_each(mbed_stats_thread_t *stats, size_t count);
+
+```
+
+RAM requirement `sizeof(SocketAddress) + sizeof(Socket*) + sizeof(enum) + sizeof(uint32_t)*3=26+4+4+4*3 = 46` per socket stats. How many we keep in memory? 10?
+
+
+### Network status events (For Mbed OS 5.12)
+
+Because in the future, the event handler needs API changes anyway, as requested in [ONME-3917](https://jira.arm.com/browse/ONME-3917) we can then consider adding statistics collection into status events. This would be done so that `NetworkInterface` would be a wrapper between user callbacks and callbacks coming from the driver.
+
+Status events should capture system timestamp associated with the events. API is similar to some in `mbed_stats.h` that requests array of events. Internally the stats should have some cap of how many events can be kept in memory. Perhaps configurable?
+
+```
+typedef struct {
+    uint32_t tick;                   /**< osKernelGetTick() value of when event happened, or equivalent timestamp */
```

Comment:
should this be 64bit timestamp (`Kernel::get_ms_count()`) or sufficient having 32 bit?

Reply:
"or equivalent"

;)

So the implementation can do this correctly by using `uint64_t` as that is what Mbed OS API gives.